### ISSUE TITLE
Support diagonal renaming in funsor.Tensor

### ISSUE THREAD

**Issue by eb8680** [pyro-ppl/funsor#381]
Bug discovered while updating https://github.com/pyro-ppl/pyro/pull/2614

The following raises an error:
```py
from collections import OrderedDict
import torch
import funsor
funsor.set_backend("torch")

x = funsor.Tensor(torch.randn(2, 2, 3), OrderedDict(a=funsor.Bint[2], b=funsor.Bint[2], c=funsor.Bint[3]), 'real')
y = x(a="d", b="d")  # error here, in Tensor.__init__
# should produce
assert set(y.inputs) == {"d", "c"}
```

**Comment by fritzo** [comment]
Good catch. I guess we should at least try to informatively error in the wide class of unsupported substitutions.

**Comment by eb8680** [comment]
I pushed a fix in #382, ideally we should merge this before the Pyro example lands.
